rhobs/alerting/data_plane/prometheus.pipeline_alerts.yaml

```diff
@@ -183,3 +183,75 @@ spec:
+        - alert: PipelinesControllerTaskRunRecLeasesNotDistributed
+          expr: count by (source_cluster) (count by (lease_holder, source_cluster) (kube_lease_owner{lease=~"tekton-pipelines-controller.github.com.tektoncd.pipeline.pkg.reconciler.taskrun.reconciler..*", lease_holder!=""})) <= 1
+          for: 5m
+          labels:
+            severity: warning
+          annotations:
+            summary: Tekton Pipelines controller TaskRun reconciler leases not distributed across more than one pod
+            description: >-
+              Tekton Pipelines controller TaskRun reconciler leases not distributed across more than one pod
+              on cluster {{ $labels.source_cluster }}.
+            alert_team_handle: <!subteam^S04PYECHCCU>
+            team: pipelines
+            runbook_url: https://gitlab.cee.redhat.com/konflux/docs/sop/-/blob/main/o11y/alert-rule-PipelinesControllerLeasesNotDistributed.md
+        - alert: PipelinesControllerResReqRecLeasesNotDistributed
+          expr: count by (source_cluster) (count by (lease_holder, source_cluster) (kube_lease_owner{lease=~"tekton-pipelines-controller.github.com.tektoncd.pipeline.pkg.reconciler.resolutionrequest.reconciler..*", lease_holder!=""})) <= 1
+          for: 5m
+          labels:
+            severity: warning
+          annotations:
+            summary: Tekton Pipelines controller ResolutionRequest reconciler leases not distributed across more than one pod
+            description: >-
+              Tekton Pipelines controller ResolutionRequest reconciler leases not distributed across more than one pod
+              on cluster {{ $labels.source_cluster }}.
+            alert_team_handle: <!subteam^S04PYECHCCU>
+            team: pipelines
+            runbook_url: https://gitlab.cee.redhat.com/konflux/docs/sop/-/blob/main/o11y/alert-rule-PipelinesControllerLeasesNotDistributed.md
+        - alert: PipelinesControllerPipelineRunRecLeasesMetricMissing
+          expr: absent(kube_lease_owner{lease=~"tekton-pipelines-controller.github.com.tektoncd.pipeline.pkg.reconciler.pipelinerun.reconciler..*", lease_holder!=""}) == 1
+          for: 5m
+          labels:
+            severity: warning
+          annotations:
+            summary: Tekton Pipelines controller PipelineRun reconciler leases metric is missing
+            description: Tekton Pipelines controller PipelineRun reconciler leases metric is missing or lease_holder is empty.
+            alert_team_handle: <!subteam^S04PYECHCCU>
+            team: pipelines
+            runbook_url: https://gitlab.cee.redhat.com/konflux/docs/sop/-/blob/main/o11y/alert-rule-PipelinesControllerLeasesNotDistributed.md
+        - alert: PipelinesControllerTaskRunRecLeasesMetricMissing
+          expr: absent(kube_lease_owner{lease=~"tekton-pipelines-controller.github.com.tektoncd.pipeline.pkg.reconciler.taskrun.reconciler..*", lease_holder!=""}) == 1
```

Comment:
Does this value of `lease` only exist in one cluster? Suppose that it exists in more than one, this alert will fire only when it's missing from both. No? I suggest adding a test case with more than a singled `source_cluster` for this alert as well.

Reply:
Hello!

These leases are there in all the clusters, so you are right! If I add `absent(sum by (source_cluster) (...)) == 1` would it help?

Yes, will add tests with multiple `source_cluster`s.

Reply:
Hello @yftacherzog . I was thinking about it and what would you think about changing the rule to:
```
(count by (source_cluster) (kube_lease_owner{lease=~"tekton-pipelines-controller.github.com.tektoncd.pipeline.pkg.reconciler.pipelinerun.reconciler..*", lease_holder!=""}) or on() vector(0)) == 0
```
?

Reply:
We saw many examples using `or on() vector(0)`, but when trying to use it, we could not find a single scenario in which that worked for us.
Having said that, it might work... The important bit is to create the relevant test case that represents what you're trying to solve and see that it indeed solves it.

it might be that incorporating the [absent](https://prometheus.io/docs/prometheus/latest/querying/functions/#absent) query function into this rule (or another complementing rule) can simplify the whole thing.

Reply:
Thank you. I was not able to figure out how to use `absent(...)` to check metric on cluster A is missing while it is still there on cluster B. Will take a look into it again.

Reply:
What is it that you're trying to alert on in this case?

Reply:
I want to ensure that lease_holder is always set:
```
kube_lease_owner{lease=~"tekton-pipelines-controller.github.com.tektoncd.pipeline.pkg.reconciler.pipelinerun.reconciler..*", lease_holder!=""}
```
Basically that this ^^^ is not absent.

Reply:
It might be that [absent_over_time](https://prometheus.io/docs/prometheus/latest/querying/functions/#absent_over_time) is a better fit here.

The reason being that the metric only gives a value of 1 when the state it represents happens, and stops providing values when it's not the case anymore, so I believe that as long as the data series exists, it will not be absent.

Reply:
In any case, you might benefit from dealing with the different alerts you're working on here on separate PRs. If they don't strictly depend on each other then it might be that you can have some of them merged while figuring out the others.

Reply:
Hello! Exactly. Just submitted simplified version. I gave up on the "metric is absent" bit of the alert and test (as that actually should not happen and if it does, it probably means whole controller is down), and only stick to "lease_holder is not empty". It is far easier to create rule and test for it now. Please take a look.